Jenkinsfile

```diff
@@ -872,6 +903,104 @@ def isChangedXPackCode(patterns) {
   return isChanged(allPatterns)
 }
 
+// withCloudTestEnv executes a closure with credentials for cloud test
+// environments.
+def withCloudTestEnv(Closure body) {
+  def maskedVars = []
+  def testTags = "${env.TEST_TAGS}"
+
+  // AWS
+  if (params.allCloudTests || params.awsCloudTests) {
+    testTags = "${testTags},aws"
+    def aws = getVaultSecret(secret: "${AWS_ACCOUNT_SECRET}").data
+    if (!aws.containsKey('access_key')) {
+      error("${AWS_ACCOUNT_SECRET} doesn't contain 'access_key'")
+    }
+    if (!aws.containsKey('secret_key')) {
+      error("${AWS_ACCOUNT_SECRET} doesn't contain 'secret_key'")
+    }
+    maskedVars.addAll([
+      [var: "AWS_REGION", password: params.awsRegion],
+      [var: "AWS_ACCESS_KEY_ID", password: aws.access_key],
+      [var: "AWS_SECRET_ACCESS_KEY", password: aws.secret_key],
+    ])
+  }
+
+  withEnv([
+    "TEST_TAGS=${testTags}",
+  ]) {
+    withEnvMask(vars: maskedVars) {
+      body()
+    }
+  }
+}
+
+def terraformInit(String directory) {
+  dir(directory) {
+    sh(label: "Terraform Init on ${directory}", script: "terraform init")
+  }
+}
+
+def terraformApply(String directory) {
+  terraformInit(directory)
+  dir(directory) {
+    sh(label: "Terraform Apply on ${directory}", script: "terraform apply -auto-approve")
+  }
+}
+
+def terraformStash(String name) {
+  archiveArtifacts(allowEmptyArchive: true, artifacts: '**/terraform.tfstate')
+  stash(name: "terraform-${name}", allowEmpty: true, includes: '**/terraform.tfstate,**/.terraform/**')
+}
+
+
+// Start testing environment on cloud using terraform and stash its state
+// for manual cleanup if needed.
+//
+// Example:
+//   startCloudTestEnv('x-pack-metricbeat', [
+//     [cond: params.awsCloudTests, dir: 'x-pack/metricbeat/module/aws'],
+//   ])
+//   ...
+//   terraformCleanup('x-pack-metricbeat', 'x-pack/metricbeat')
+def startCloudTestEnv(String name, environments = []) {
+  withCloudTestEnv() {
+    withBeatsEnv(false) {
+      script {
+        def runAll = params.runAllCloudTests
+        def failed = false
+        for (environment in environments) {
+          if (environment.cond || runAll) {
+            try {
+              terraformApply(environment.dir)
+            } catch (Exception e) {
+              failed = true
+            }
+          }
+        }
+        terraformStash(name)
+        if (failed) {
+          error("Some environment failed to start")
+        }
```

Comment:
why execute all the environment deploys even do one failed, to fail in case one fail, e.g 

enviroments=[ A, B ,C]
it will call  terraformApply for A, B,  and C
let's say that B fails, the failed var is set to false and we call terraformApply(C) and success.
we stash the terraform status
and finally, we throw an error because the failed var is true

Reply:
Good point. I tried to mark it as unstable but let it continue so other tests are executed, but at the end I left it half way there. In any case this is only executed if someone wants to test cloud integrations, so it should fail in any case. I am going to modify it so it stops as soon as something fails.

Reply:
Done in 04c7292.